framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -240,48 +246,50 @@ private AnnotationMirror createBooleanAnnotationMirror(List<Boolean> values) {
             Node rightNode,
             NumericalBinaryOps op,
             TransferInput<CFValue, CFStore> p) {
-        List<? extends Number> lefts = getNumericalValues(leftNode, p);
-        List<? extends Number> rights = getNumericalValues(rightNode, p);
         List<Number> resultValues = new ArrayList<>();
-        for (Number left : lefts) {
-            NumberMath<?> nmLeft = NumberMath.getNumberMath(left);
-            for (Number right : rights) {
-                switch (op) {
-                    case ADDITION:
-                        resultValues.add(nmLeft.plus(right));
-                        break;
-                    case DIVISION:
-                        resultValues.add(nmLeft.divide(right));
-                        break;
-                    case MULPLICATION:
-                        resultValues.add(nmLeft.times(right));
-                        break;
-                    case REMAINDER:
-                        resultValues.add(nmLeft.remainder(right));
-                        break;
-                    case SUBTRACTION:
-                        resultValues.add(nmLeft.minus(right));
-                        break;
-                    case SHIFT_LEFT:
-                        resultValues.add(nmLeft.shiftLeft(right));
-                        break;
-                    case SIGNED_SHIFT_RIGHT:
-                        resultValues.add(nmLeft.signedSiftRight(right));
-                        break;
-                    case UNSIGNED_SHIFT_RIGHT:
-                        resultValues.add(nmLeft.unsignedSiftRight(right));
-                        break;
-                    case BITWISE_AND:
-                        resultValues.add(nmLeft.bitwiseAnd(right));
-                        break;
-                    case BITWISE_OR:
-                        resultValues.add(nmLeft.bitwiseOr(right));
-                        break;
-                    case BITWISE_XOR:
-                        resultValues.add(nmLeft.bitwiseXor(right));
-                        break;
-                    default:
-                        throw new UnsupportedOperationException();
+        if (!isIntRange(leftNode, p) && !isIntRange(rightNode, p)) {
```

Comment:
Thanks for the reply -- I appreciate it.

I agree with you that we want to trade off precision with complexity.  In this pull request, we should not implement something that is extremely complicated, especially if we don't know that it is valuable.  But if something is easy to do, we should consider it.

Here is a proposal:  if one annotation is `@IntVal` and the other is `@IntRange`, convert the `@IntVal` to `@IntRange` and then do the operation.  This has these advantages:
 * easy to write and verify
 * don't need to create any new rules for operations with mixed types
 * no tricky special cases in the translation among annotations, since `@IntVal` can always be represented as `@IntRange`
 * probably precise enough in practice; if it isn't precise enough, then a future issue and pull request can deal with it.

What do you think of that?